### PR TITLE
Collective cleanup

### PR DESCRIPTION
**Pull request by pavanbalaji**
This depends on PR #2859 and is a follow-up set of patches.

### CODE REVIEW

**Comment by raffenet** [comment]
@pavanbalaji is this ready to go? I can review it this morning.
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
There are two more things left to be done in this PR:

1. The neighborhood collectives do not have intra/inter functions, that need to be added.

2. The CVAR selection currently happens in the `MPIR_<coll>` functions instead of the `MPIR_<coll>_intra` and `MPIR_<coll>_inter` functions.  This means that if the device or a different collective uses `MPIR_<coll>_intra` directly, the user-set CVAR has no impact on the algorithm that is chosen.  So, in some sense, the `MPIR_<coll>_intra` and `MPIR_<coll>_inter` functions behave as if they are the `MPIR_<coll>_intra_auto` and `MPIR_<coll>_inter_auto` algorithms.  They should either be renamed to be prepended with `_auto` or new functions called `MPIR_<coll>_intra_auto` and `MPIR_<coll>_inter_auto` should be added (in the latter case, `MPIR_<coll>_intra` and `MPIR_<coll>_inter` should simply pick the algorithm set by CVARs).

File: src/mpi/romio/adio/common/flatten.c [hunk 1]
```diff
@@ -504,7 +504,7 @@ void ADIOI_Flatten(MPI_Datatype datatype, ADIOI_Flatlist_node *flat,
     case MPI_COMBINER_HINDEXED_BLOCK:
 	is_hindexed_block=1;
 	/* deliberate fall-through */
-    MPL_FALLTHROUGH;
+	MPL_FALLTHROUGH;
```
Comment (wesbland) [inline src/mpi/romio/adio/common/flatten.c]: Probably not worth doing this before #2845, but I don't care that much.

File: src/include/mpir_coll.h [hunk 2]
```diff
@@ -312,7 +312,7 @@ int MPIR_Reduce_scatter_block_nb(const void *sendbuf, void *recvbuf, int recvcou
 /* Reduce functions */
 int MPIR_Reduce(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
                 MPI_Op op, int root, MPIR_Comm *comm_ptr, MPIR_Errflag_t *errflag );
-int MPIR_Reduce_intra_redscat_gather(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
+int MPIR_Reduce_intra_reduce_scatter_gather(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
```
Comment (wesbland) [inline src/include/mpir_coll.h]: There's still a bunch of places in CH4 that use redscat:

```
src/mpid/ch4/netmod/ofi/ofi_coll_containers.h:14:extern const MPIDI_OFI_coll_algo_container_t OFI_reduce_redscat_gather_cnt;
src/mpid/ch4/netmod/ofi/ofi_coll_globals_default.c:24:const MPIDI_OFI_coll_algo_container_t OFI_reduce_redscat_gather_cnt =
src/mpid/ch4/netmod/ofi/ofi_coll_globals_default.c:26:        .id = MPIDI_OFI_reduce_redscat_gather_id
src/mpid/ch4/netmod/ofi/ofi_coll_impl.h:127:#define FUNCNAME MPIDI_OFI_reduce_redscat_gather
src/mpid/ch4/netmod/ofi/ofi_coll_impl.h:131:int MPIDI_OFI_reduce_redscat_gather(const void *sendbuf, void *recvbuf, int count,
src/mpid/ch4/netmod/ofi/ofi_coll_params.h:45:    MPIDI_OFI_reduce_redscat_gather_id,
src/mpid/ch4/netmod/ofi/ofi_coll_select.h:161:        return (MPIDI_OFI_coll_algo_container_t *) &OFI_reduce_redscat_gather_cnt;
src/mpid/ch4/netmod/ofi/ofi_coll_select.h:176:    case MPIDI_OFI_reduce_redscat_gather_id:
src/mpid/ch4/netmod/ofi/ofi_coll_select.h:178:            MPIDI_OFI_reduce_redscat_gather(sendbuf, recvbuf, count, datatype, op, root, comm_ptr,
src/mpid/ch4/netmod/portals4/ptl_coll_globals_default.c:24:const MPIDI_PTL_coll_algo_container_t PTL_reduce_redscat_gather_cnt =
src/mpid/ch4/netmod/portals4/ptl_coll_globals_default.c:26:        .id = MPIDI_PTL_reduce_redscat_gather_id
src/mpid/ch4/netmod/portals4/ptl_coll_params.h:45:    MPIDI_PTL_reduce_redscat_gather_id,
src/mpid/ch4/netmod/stubnm/stubnm_coll_globals_default.c:24:const MPIDI_STUBNM_coll_algo_container_t STUBNM_reduce_redscat_gather_cnt =
src/mpid/ch4/netmod/stubnm/stubnm_coll_globals_default.c:26:        .id = MPIDI_STUBNM_reduce_redscat_gather_id
src/mpid/ch4/netmod/stubnm/stubnm_coll_params.h:45:    MPIDI_STUBNM_reduce_redscat_gather_id,
src/mpid/ch4/netmod/ucx/ucx_coll_globals_default.c:24:const MPIDI_UCX_coll_algo_container_t UCX_reduce_redscat_gather_cnt =
src/mpid/ch4/netmod/ucx/ucx_coll_globals_default.c:26:        .id = MPIDI_UCX_reduce_redscat_gather_id
src/mpid/ch4/netmod/ucx/ucx_coll_params.h:45:    MPIDI_UCX_reduce_redscat_gather_id,
src/mpid/ch4/shm/posix/posix_coll_containers.h:14:extern const MPIDI_POSIX_coll_algo_container_t POSIX_reduce_redscat_gather_cnt;
src/mpid/ch4/shm/posix/posix_coll_globals_default.c:24:const MPIDI_POSIX_coll_algo_container_t POSIX_reduce_redscat_gather_cnt =
src/mpid/ch4/shm/posix/posix_coll_globals_default.c:26:        .id = MPIDI_POSIX_reduce_redscat_gather_id
src/mpid/ch4/shm/posix/posix_coll_impl.h:127:#define FUNCNAME MPIDI_POSIX_reduce_redscat_gather
src/mpid/ch4/shm/posix/posix_coll_impl.h:131:int MPIDI_POSIX_reduce_redscat_gather(const void *sendbuf, void *recvbuf, int count,
src/mpid/ch4/shm/posix/posix_coll_params.h:33:    MPIDI_POSIX_reduce_redscat_gather_id,
src/mpid/ch4/shm/posix/posix_coll_select.h:167:        return (MPIDI_POSIX_coll_algo_container_t *) &POSIX_reduce_redscat_gather_cnt;
src/mpid/ch4/shm/posix/posix_coll_select.h:182:    case MPIDI_POSIX_reduce_redscat_gather_id:
src/mpid/ch4/shm/posix/posix_coll_select.h:184:            MPIDI_POSIX_reduce_redscat_gather(sendbuf, recvbuf, count, datatype, op, root, comm_ptr,
src/mpid/ch4/shm/stubshm/stubshm_coll_globals_default.c:24:const MPIDI_STUBSHM_coll_algo_container_t STUBSHM_reduce_redscat_gather_cnt =
src/mpid/ch4/shm/stubshm/stubshm_coll_globals_default.c:26:        .id = MPIDI_STUBSHM_reduce_redscat_gather_id
src/mpid/ch4/shm/stubshm/stubshm_coll_params.h:33:    MPIDI_STUBSHM_reduce_redscat_gather_id,
```

File: src/mpi/coll/include/coll_impl.h [hunk 3]
```diff
@@ -178,41 +178,41 @@ typedef enum MPIR_Iallgather_intra_algo_t {
     MPIR_IALLGATHER_INTRA_ALGO_AUTO,
     MPIR_IALLGATHER_INTRA_ALGO_BRUCKS,
     MPIR_IALLGATHER_INTRA_ALGO_RECURSIVE_DOUBLING,
-    MPIR_IALLGATHER_INTRA_ALGO_RING
+    MPIR_IALLGATHER_INTRA_ALGO_RING,
```
Comment (wesbland) [inline src/mpi/coll/include/coll_impl.h]: I honestly don't care enough to move this kind of thing to separate files, but it probably should be mentioned as a second cleanup in the commit message.

File: src/mpi/coll/src/coll_impl.c [hunk 4]
```diff
@@ -311,19 +311,19 @@ int MPIR_COLL_init(void) {
 
     /* Gatherv Intra */
     if (0 == strcmp(MPIR_CVAR_GATHERV_INTRA_ALGORITHM, "linear"))
-        MPIR_Gather_intra_algo_choice = MPIR_GATHERV_INTRA_ALGO_LINEAR;
+        MPIR_Gatherv_intra_algo_choice = MPIR_GATHERV_INTRA_ALGO_LINEAR;
```
Comment (wesbland) [inline src/mpi/coll/src/coll_impl.c]: I think this is unrelated to the rest of the patch.

Comment (wesbland) [reply src/mpi/coll/src/coll_impl.c]: Maybe it goes in the next one?

File: src/mpid/common/sched/mpidu_sched.c [hunk 5]
```diff
@@ -490,13 +488,9 @@ int MPIDU_Sched_start(MPIR_Sched_t * sp, MPIR_Comm * comm, int tag, MPIR_Request
 
     /* finally, enqueue in the list of all pending schedules so that the
      * progress engine can make progress on it */
-    if (all_schedules.head == NULL) {
-        mpi_errno = MPID_Progress_register_hook(MPIDU_Sched_progress, &nbc_progress_hook_id);
-        if (mpi_errno)
-            MPIR_ERR_POP(mpi_errno);
+    if (all_schedules.head == NULL)
```
Comment (wesbland) [inline src/mpid/common/sched/mpidu_sched.c]: Is there any chance we have to worry about thread safety here?

**Comment by wesbland** [comment]
> The CVAR selection currently happens in the MPIR_<coll> functions instead of the MPIR_<coll>_intra and MPIR_<coll>_inter functions. This means that if the device or a different collective uses MPIR_<coll>_intra directly, the user-set CVAR has no impact on the algorithm that is chosen. So, in some sense, the MPIR_<coll>_intra and MPIR_<coll>_inter functions behave as if they are the MPIR_<coll>_intra_auto and MPIR_<coll>_inter_auto algorithms. They should either be renamed to be prepended with _auto or new functions called MPIR_<coll>_intra_auto and MPIR_<coll>_inter_auto should be added (in the latter case, MPIR_<coll>_intra and MPIR_<coll>_inter should simply pick the algorithm set by CVARs).

I vote for the first option (append `_auto` to the function name). Otherwise, it looks like we have to do algorithm selection in two places.

**Comment by pavanbalaji** [comment]
Why will it be in two places?  `MPIR_<coll>_intra` would do the switch statement with the different CVARs, while `MPIR_<coll>_intra_auto` will pick one or more of the algorithms based on message size/comm size.  `MPIR_<coll>` would simply call `MPIR_<coll>_intra` or `MPIR_<coll>_inter`.

**Comment by wesbland** [comment]
Oh, you want to move it all down from `MPIR_<coll>` to `MPIR_<coll>_intra`? So the `MPIR_<coll>` function would just pick intra/inter? That'd be fine too. The elegance of the current system is that all of the selection code is in one place (intra/inter and algorithm).

Another alternative is to change the other uses of `MPIR_<coll>_intra` to use `MPIR_<coll>` instead to keep everything in one place. The downside of this one is that we'd add a branch for those cases.

**Comment by pavanbalaji** [comment]
Either moving everything into `MPIR_<coll>intra` or completely removing the intra/inter functions is fine with me (even though the latter adds an extra branch).  Do you have a preference?

File: src/mpid/common/sched/mpidu_sched.c (hunk 5 above)
Comment (pavanbalaji) [reply src/mpid/common/sched/mpidu_sched.c]: Not for the global lock case.  Even in the workq model that we are working on, the progress engine is always protected with a lock.  So, I believe, this will be correct either way.

**Comment by wesbland** [comment]
I think I’d prefer removing the functions.

File: src/mpid/common/sched/mpidu_sched.c (hunk 5 above)
Comment (wesbland) [reply src/mpid/common/sched/mpidu_sched.c]: Ok. That sounds good then.

**Comment by pavanbalaji** [comment]
So, is the plan to have an `_auto` function?  Otherwise the switch statement can become too cluttered.

**Comment by wesbland** [comment]
Yeah. I think we’d have to.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
@akhillanger your patches look fine.  I have added ifdef error checking in one of the patches, so the error checks don't interfere with the fast path.  Please take a look at it.  I'll be pushing my changes once I test them on my laptop.  I'll kickstart the tests again after that.

I'm done with my set of changes.  Do you have any other changes that need to go into the prereqs?

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
The tests are failing with some weird jenkins issue.  This might need to wait till @raffenet or @yfguo can look at the jenkins issues after the break.  Apart from that, the branch is ready for review.

@wesbland I have incorporated your comments above.

@akhillanger I have signed off on your patches.

**Comment by akhillanger** [comment]
Thanks @pavanbalaji for adding the error checking macros. They look fine. With that, I had to further edit that commit to remove some unused vars. I will have a couple of more commits for the cleanup branch. But I agree that we should merge the ones that have been signed-off, to make it easy to track.

Some of the algos already had error checks but are not protected by the error checking macros. Do you want me to add the error checking macro to those as a separate commit or in the existing commit?

**Comment by pavanbalaji** [comment]
Can you remove my signoffs from the patches that you edited, so I know I need to look at them again?

**Comment by akhillanger** [comment]
Sure. Just one patch.

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
All the tests look good. The ofi failures are unrelated.

**Comment by akhillanger** [comment]
There is a warning in the tests. I am looking at it.

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
There was a build issue with one of the configs. That is fixed now. Tests should be good now.

**Comment by akhillanger** [comment]
All tests results look good. I think we should start merging the commits, starting with the ones that have been signed-off.

File: src/mpi/coll/ibcast/ibcast.c [hunk 6]
```diff
@@ -108,11 +108,11 @@ int MPIR_Ibcast_sched_intra(void *buffer, int count, MPI_Datatype datatype, int
     else /* (nbytes >= MPIR_CVAR_BCAST_SHORT_MSG_SIZE) && (comm_size >= MPIR_CVAR_BCAST_MIN_PROCS) */
     {
         if ((nbytes < MPIR_CVAR_BCAST_LONG_MSG_SIZE) && (MPIU_is_pof2(comm_size, NULL))) {
-            mpi_errno = MPII_Ibcast_scatter_rec_dbl_allgather_sched(buffer, count, datatype, root, comm_ptr, s);
+            mpi_errno = MPIR_Ibcast_sched_intra_scatter_recursive_doubling_allgather(buffer, count, datatype, root, comm_ptr, s);
```
Comment (wesbland) [inline src/mpi/coll/ibcast/ibcast.c]: If we're going to convert these to the namespace, should they not also be added to the selection algorithm (CVAR, enum, etc.)?

File: src/include/mpir_coll.h [hunk 7]
```diff
@@ -1396,14 +1402,16 @@ int MPIR_Scatterv(const void *sendbuf, const int *sendcnts, const int *displs,
...
+                             MPIR_Errflag_t * errflag);
 
 /* intercomm-only functions */
-int MPIR_Scatterv_inter(const void *sendbuf, const int *sendcnts, const int *displs,
-                        MPI_Datatype sendtype, void *recvbuf, int recvcnt, MPI_Datatype recvtype,
-                        int root, MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag);
+int MPIR_Scatterv_inter_auto(const void *sendbuf, const int *sendcnts, const int *displs,
+                             MPI_Datatype sendtype, void *recvbuf, int recvcnt,
+                             MPI_Datatype recvtype, int root, MPIR_Comm * comm_ptr,
+                             MPIR_Errflag_t * errflag);
```
Comment (wesbland) [inline src/include/mpir_coll.h]: This change skips a lot of collectives, such as all of the non-blocking ones.

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
The ch4 tests, which were failing earlier, are looking good now.

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch4/ucx

**Comment by akhillanger** [comment]
ucx tests are passing. Starting full tests now.

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch4/ofi

**Comment by akhillanger** [comment]
test:jenkins/ch4/ofi

**Comment by akhillanger** [comment]
test:jenkins/ch4/ofi

**Comment by akhillanger** [comment]
test:jenkins/ch4/ofi

**Comment by akhillanger** [comment]
ch4-ofi tests are also passing now.

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
@pavanbalaji all checks have passed. Once you have seen this, I will squash the last fixup commit (that is just a typo fix) and then we can merge this pull request. Let me know.

**Comment by pavanbalaji** [comment]
Looks good.  Signedoff and merging.
